Zend/zend_API.c

```diff
@@ -3723,58 +3724,151 @@ ZEND_API int zend_declare_property_stringl(zend_class_entry *ce, const char *nam
 }
 /* }}} */
 
-ZEND_API int zend_declare_class_constant(zend_class_entry *ce, const char *name, size_t name_length, zval *value) /* {{{ */
+ZEND_API int zend_declare_class_constant_ex(zend_class_entry *ce, zend_string *name, zval *value, int access_type, zend_string *doc_comment) /* {{{ */
 {
+	zend_class_constant_info *const_info;
+
+	if (zend_string_equals_literal_ci(name, "class")) {
+		zend_error(E_CORE_ERROR,
+				"A class constant must not be called 'class'; it is reserved for class name fetching");
```

Comment:
Note: Moving the check in here changes the error from E_COMPILE_ERROR to E_CORE_ERROR. I'd probably choose between these to depending on whether it's an internal class or not.